source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc

```diff
@@ -0,0 +1,190 @@
+  // Set the minRTT flag to indicate we're gathering samples to update the value. This will
+  // prevent the sample window from resetting until enough requests are gathered to complete the
+  // recalculation.
+  concurrency_limit_.store(1);
+  stats_.concurrency_limit_.set(concurrency_limit_.load());
+  recalculating_min_rtt_.store(true);
+
+  // Throw away any latency samples from before the recalculation window as it may not represent
+  // the minRTT.
+  absl::MutexLock ml(&latency_sample_mtx_);
+  hist_clear(latency_sample_hist_);
+}
+
+void GradientController::updateMinRTT() {
+  ASSERT(recalculating_min_rtt_.load());
+
+  // Reset the timer to ensure the next minRTT sampling window upon leaving scope.
+  auto defer =
+      Cleanup([this]() { min_rtt_calc_timer_->enableTimer(config_->min_rtt_calc_interval()); });
+
+  absl::MutexLock ml(&latency_sample_mtx_);
+  min_rtt_ = processLatencySamplesAndClear();
+  stats_.min_rtt_msecs_.set(
+      std::chrono::duration_cast<std::chrono::milliseconds>(min_rtt_).count());
+  recalculating_min_rtt_.store(false);
+}
+
+void GradientController::resetSampleWindow() {
+  // Reset the timer upon leaving scope.
+  auto defer =
+      Cleanup([this]() { sample_reset_timer_->enableTimer(config_->sample_rtt_calc_interval()); });
+
+  // The sampling window must not be reset while sampling for the new minRTT value.
+  if (recalculating_min_rtt_.load()) {
+    return;
+  }
+
+  absl::MutexLock ml(&latency_sample_mtx_);
+  if (hist_sample_count(latency_sample_hist_) == 0) {
+    return;
+  }
+
+  sample_rtt_ = processLatencySamplesAndClear();
+  concurrency_limit_.store(calculateNewLimit());
+  stats_.concurrency_limit_.set(concurrency_limit_.load());
+}
+
+std::chrono::microseconds GradientController::processLatencySamplesAndClear() {
+  const std::array<double, 1> quantile{config_->sample_aggregate_percentile()};
+  std::array<double, 1> ans;
+  hist_approx_quantile(latency_sample_hist_, quantile.data(), 1, ans.data());
+  hist_clear(latency_sample_hist_);
+  return std::chrono::microseconds(static_cast<int>(ans[0]));
+}
+
+int GradientController::calculateNewLimit() {
+  const double gradient =
+      std::min(config_->max_gradient(), double(min_rtt_.count()) / sample_rtt_.count());
+  stats_.gradient_.set(gradient);
+  const double limit = concurrency_limit_.load() * gradient;
+  const double burst_headroom = sqrt(limit);
+  stats_.burst_queue_size_.set(burst_headroom);
+
+  const auto clamp = [](int min, int max, int val) { return std::max(min, std::min(max, val)); };
+  return clamp(1, config_->max_concurrency_limit(), static_cast<int>(limit + burst_headroom));
+}
+
+RequestForwardingAction GradientController::forwardingDecision() {
+  while (true) {
+    int curr_outstanding = num_rq_outstanding_.load();
+    if (curr_outstanding < concurrency_limit_.load()) {
+      // Using the weak compare/exchange for improved performance at the cost of spurious failures.
```

Comment:
I think for now we can just do away with this extra complexity, it's OK to have an eventually consistent answer. If you want to put in a TODO for reconsideration layer no problem. Then the logic above can be simplified to a direct compare.